### ISSUE TITLE
"let" not supported in right hand side of rewrite rule

### ISSUE THREAD

**Issue by gabrielhdt** [Deducteam/lambdapi#215]
A rule of the form
```
rule mult (S &x) &y → let larg = &y in plus larg (mult &x larg)
```
fails raising `Only allowed in terms` although the rhs is a valid term according to `syntax.bnf`.

**Comment by rlepigre** [comment]
That's not a bug, but a feature. This is linked to the fact that annotations on abstractions are mandatory in the RHS (and `let` is compiled down to a redex).

Of course, we can discuss whether having `let` bindings in RHS is useful in practice. My personal feeling about that is: they are not very useful.

**Comment by gabrielhdt** [comment]
They might be useful to introduce sharing in a non linear rhs of a rewrite rule. If I compute `mult x y` and `y` is, say, 54!, I'd like the 54! to be computed only once and not twice.  I thought that doing a let as mentioned above would allow one to compute the 54! only once.

**Comment by rlepigre** [comment]
No, not at all. For now, `let x = t in u` is exactly the same thing as `(λx.u) t`. This might change in the future, but this has implications in the theory so we cannot just add a primitive `let`.

**Comment by fblanqui** [comment]
I don't see your point Rodolphe. If t is 54!, it will be computed only once when reducing (λx.u) t.

Other remarks:
- Gabriel is right: the doc or error message is not clear enough.
- You are right that there is problem because let's have no type annotation and, in RHS, there cannot be metavariables. But this could be fixed. This is the object of https://github.com/Deducteam/lambdapi/issues/169.

**Comment by rlepigre** [comment]
@fblanqui: no, it may be computed any number of times since we are not doing call-by-value evaluation (nor a fully lazy evaluation). There is only some sharing in the case where rewriting rules are involved, but again, not always. At some point Christophe experimented with lazy evaluation, but there was not a massive gain in performances. However, that is something we could consider again at some point.

**Comment by rlepigre** [comment]
I just pushed some improvements to error messages (an positions) in scoping. Now the error should be clear. Now that we have fixed #169 we can choose to allow let-bindings in a RHS, but I don't think this is useful. Any thoughts on that @fblanqui? (Note that my comment above still holds.)

**Comment by gabrielhdt** [comment]
Allowing `let` in rhs seems useful to me, see:
```
rule times (frac &a &b) (frac &c &d) →
  let bv ≔ fst &b in
  let bpr ≔ snd &b in
  let dv ≔ fst &d in
  let dpr ≔ snd &d in
  frac
   (N.times &a &c)
   (N.nznatc (N.times bv dv) (N.prod_not_zero bv dv bpr dpr))
```
(I am not sure it would typecheck, but I'd like it to...)

**Comment by gabrielhdt** [comment]
fixed in fda8752
